File: src/java/product-pricing/src/main/java/com/product/pricing/core/PricingService.java
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2024 Datadog, Inc.
 */

package com.product.pricing.core;

import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.concurrent.TimeUnit;

@Service
public class PricingService {
    private final EventPublisher eventPublisher;

    public PricingService(EventPublisher eventPublisher) {
        this.eventPublisher = eventPublisher;
    }
    
    public void calculatePricing(String productId, Double currentPrice) {
        // Force a Lambda timeout
        if (currentPrice > 50 && currentPrice < 60) {
            try{
                TimeUnit.SECONDS.sleep(40);
            }
            catch (InterruptedException e){}
        }
        
        // Force an error
        if (currentPrice > 90 && currentPrice < 95) {
            throw new ProductPricingError();
        }
        
        HashMap<Double, Double> pricingBreakdown = new HashMap<>();
        pricingBreakdown.put(5.0, currentPrice * 0.95);
        pricingBreakdown.put(10.0, currentPrice * 0.9);
        pricingBreakdown.put(25.0, currentPrice * 0.8);
        pricingBreakdown.put(50.0, currentPrice * 0.75);
        pricingBreakdown.put(100.0, currentPrice * 0.7);
        
        this.eventPublisher.publishPriceCalculatedEvent(new ProductPriceCalculatedEvent(productId, pricingBreakdown));
    }
}

File: src/order-service/src/Product.Pricing/ProductPricingService.Core/PricingService.cs
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2024 Datadog, Inc.

namespace ProductPricingService.Core;

public class PricingService(IEventPublisher eventPublisher)
{
    public async Task GeneratePricingFor(string productId, ProductPrice price)
    {
        if (price.Value > 50 && price.Value < 60){
            await Task.Delay(TimeSpan.FromSeconds(40));
        }

        if (price.Value > 90 && price.Value < 95) {
            throw new Exception("Failure in product pricing service");
        }

        var pricingOptions = new Dictionary<int, decimal>(5)
        {
            { 5, price.Value * 0.95M },
            { 10, price.Value * 0.9M },
            { 25, price.Value * 0.8M },
            { 50, price.Value * 0.75M },
            { 100, price.Value * 0.7M }
        };

        await eventPublisher.Publish(new ProductPricingUpdatedEvent(productId, pricingOptions));
    } 
}